File: js/tester.js
#!/usr/bin/env -S node --experimental-wasm-type-reflection
/**
 * tester.js - JavaScript host for the feather test harness.
 * Mirrors cmd/feather-tester/main.go definitions.
 */

import { createFeather, TCL_PARSE_OK, TCL_PARSE_INCOMPLETE, TCL_PARSE_ERROR } from './feather.js';
import { createInterface } from 'readline';
import { readFileSync, writeSync, fstatSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

async function main() {
  // Check for benchmark mode
  if (process.argv.length > 2 && process.argv[2] === '--benchmark') {
    await runBenchmarkMode();
    return;
  }

  const wasmPath = join(__dirname, 'feather.wasm');
  const feather = await createFeather(wasmPath);
  const interp = feather.create();

  registerTestCommands(feather, interp);

  // Check if stdin is a TTY
  try {
    const stat = fstatSync(0);
    if (stat.isCharacterDevice()) {
      await runREPL(feather, interp);
      return;
    }
  } catch {
    // Not a TTY, proceed with script mode
  }

  await runScript(feather, interp);
}

function registerTestCommands(feather, interp) {
  // Set milestone variables
  feather.eval(interp, 'set milestone m1');
  feather.eval(interp, 'set current-step m1');

  // Test commands
  feather.register(interp, 'say-hello', () => {
    console.log('hello');
    return '';
  });

  feather.register(interp, 'echo', (args) => {
    console.log(args.join(' '));
    return '';
  });

  feather.register(interp, 'count', (args) => {
    return String(args.length);
  });

  feather.register(interp, 'list', (args) => {
    const parts = args.map(s => {
      if (s.length === 0) return '{}';
      if (/[\s{}]/.test(s)) return '{' + s + '}';
      return s;
    });
    return parts.join(' ');
  });

  // Register the Counter foreign type
  let nextCounterId = 1;
  const counters = new Map();

  feather.registerType(interp, 'Counter', {
    methods: {
      get: (c) => c.value,
      set: (c, val) => { c.value = parseInt(val, 10); return ''; },
      incr: (c) => ++c.value,
      add: (c, amount) => { c.value += parseInt(amount, 10); return c.value; },
    },
    destroy: (c) => { counters.delete(c.id); }
  });

  // Register Counter constructor command
  feather.register(interp, 'Counter', (args) => {
    if (args.length === 0) {
      throw new Error('wrong # args: should be "Counter subcommand ?arg ...?"');
    }
    if (args[0] !== 'new') {
      throw new Error(`unknown subcommand "${args[0]}": must be new`);
    }
    const id = nextCounterId++;
    const counter = { id, value: 0 };
    counters.set(id, counter);

    const handle = `counter${id}`;
    const foreignHandle = feather.createForeign(interp, 'Counter', counter, handle);

    // Define method signatures: [expectedArgCount, needsIntArg]
    const methodDefs = {
      get: { argc: 0, fn: () => counter.value },
      set: { argc: 1, intArg: 1, fn: (val) => { counter.value = val; return ''; } },
      incr: { argc: 0, fn: () => ++counter.value },
      add: { argc: 1, intArg: 1, fn: (amount) => { counter.value += amount; return counter.value; } },
      destroy: { argc: 0, fn: () => {
        counters.delete(id);
        feather.destroyForeign(interp, handle);
        feather.register(interp, handle, () => { throw new Error(`invalid command name "${handle}"`); });
        return '';
      }}
    };

    // Register the object-as-command
    feather.register(interp, handle, (methodArgs) => {
      if (methodArgs.length === 0) {
        throw new Error(`wrong # args: should be "${handle} method ?arg ...?"`);
      }
      const method = methodArgs[0];
      const rest = methodArgs.slice(1);
      const def = methodDefs[method];
      if (!def) {
        const methodList = Object.keys(methodDefs).join(', ');
        throw new Error(`unknown method "${method}": must be ${methodList}`);
      }
      // Check argument count
      if (rest.length !== def.argc) {
        throw new Error(`wrong # args: expected ${def.argc}, got ${rest.length}`);
      }
      // Convert integer arguments if needed
      const convertedArgs = rest.map((arg, idx) => {
        if (def.intArg === idx + 1) {
          const num = parseInt(arg, 10);
          if (isNaN(num)) {
            throw new Error(`argument ${idx + 1}: expected integer but got "${arg}"`);
          }
          return num;
        }
        return arg;
      });
      return String(def.fn(...convertedArgs) ?? '');
    });

    return handle;
  });
}

async function runREPL(feather, interp) {
  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  let inputBuffer = '';

  const prompt = () => {
    rl.question(inputBuffer === '' ? '% ' : '> ', (line) => {
      if (line === undefined) {
        rl.close();
        return;
      }

      inputBuffer = inputBuffer ? inputBuffer + '\n' + line : line;

      // Simple incomplete check: count braces
      const opens = (inputBuffer.match(/{/g) || []).length;
      const closes = (inputBuffer.match(/}/g) || []).length;
      if (opens > closes) {
        prompt();
        return;
      }

      try {
        const result = feather.eval(interp, inputBuffer);
        if (result !== '') console.log(result);
      } catch (e) {
        console.error(`error: ${e.message}`);
      }
      inputBuffer = '';
      prompt();
    });
  };

  return new Promise((resolve) => {
    rl.on('close', () => {
      console.log('');
      feather.destroy(interp);
      resolve();
    });
    prompt();
  });
}

async function runScript(feather, interp) {
  const script = readFileSync(0, 'utf-8');

  // Check parse status first (like Go version)
  const parseResult = feather.parse(interp, script);
  if (parseResult.status === TCL_PARSE_INCOMPLETE) {
    writeHarnessResult('TCL_OK', parseResult.result, '');
    process.exit(2);
  }
  if (parseResult.status === TCL_PARSE_ERROR) {
    writeHarnessResult('TCL_ERROR', parseResult.result, parseResult.errorMessage);
    process.exit(3);
  }

  try {
    const result = feather.eval(interp, script);
    if (result !== '') console.log(result);
    writeHarnessResult('TCL_OK', result, '');
  } catch (e) {
    console.log(e.message);
    writeHarnessResult('TCL_ERROR', '', e.message);
    process.exit(1);
  }
}

function writeHarnessResult(returnCode, result, errorMsg) {
  if (process.env.FEATHER_IN_HARNESS !== '1') return;

  // Write to fd 3 (harness communication channel)
  let output = `return: ${returnCode}\n`;
  if (result !== '') output += `result: ${result}\n`;
  if (errorMsg !== '') output += `error: ${errorMsg}\n`;

  try {
    writeSync(3, output);
  } catch {
    // fd 3 not available, ignore
  }
}

async function runBenchmarkMode() {
  const wasmPath = join(__dirname, 'feather.wasm');
  const feather = await createFeather(wasmPath);
  const interp = feather.create();

  registerTestCommands(feather, interp);

  // Read benchmarks from stdin
  const input = readFileSync(0, 'utf-8');
  let benchmarks;
  try {
    benchmarks = JSON.parse(input);
  } catch (e) {
    console.error(`error reading benchmarks: ${e.message}`);
    process.exit(1);
  }

  // Run each benchmark
  for (const b of benchmarks) {
    const result = runSingleBenchmark(feather, interp, b);

    // Write result as JSON to fd 3
    const resultJSON = JSON.stringify(result);
    try {
      writeSync(3, resultJSON + '\n');
    } catch {
      console.error('error: harness channel not available');
      process.exit(1);
    }
  }
}

function runSingleBenchmark(feather, interp, b) {
  const result = {
    Benchmark: b,
    Success: true,
    TotalTime: 0,
    AvgTime: 0,
    MinTime: 0,
    MaxTime: 0,
    Iterations: 0,
    OpsPerSecond: 0,
    Error: ''
  };

  // Run setup if provided
  if (b.Setup) {
    try {
      feather.eval(interp, b.Setup);
    } catch (e) {
      result.Success = false;
      result.Error = `setup failed: ${e.message}`;
      return result;
    }
  }

  // Warmup iterations
  for (let w = 0; w < (b.Warmup || 0); w++) {
    try {
      feather.eval(interp, b.Script);
    } catch (e) {
      result.Success = false;
      result.Error = `warmup failed: ${e.message}`;
      return result;
    }
  }

  // Measured iterations
  const iterations = b.Iterations || 1000;
  let totalTime = 0;
  let minTime = Infinity;
  let maxTime = 0;

  for (let iter = 0; iter < iterations; iter++) {
    const start = process.hrtime.bigint();
    try {
      feather.eval(interp, b.Script);
    } catch (e) {
      result.Success = false;
      result.Error = `iteration ${iter} failed: ${e.message}`;
      return result;
    }
    const end = process.hrtime.bigint();
    const elapsed = Number(end - start); // nanoseconds

    totalTime += elapsed;
    if (elapsed < minTime) minTime = elapsed;
    if (elapsed > maxTime) maxTime = elapsed;
    result.Iterations++;
  }

  // Calculate statistics (convert from nanoseconds to time.Duration format)
  result.TotalTime = totalTime;
  result.AvgTime = Math.floor(totalTime / iterations);
  result.MinTime = minTime;
  result.MaxTime = maxTime;
  if (result.AvgTime > 0) {
    result.OpsPerSecond = 1e9 / result.AvgTime; // 1 billion ns per second
  }

  return result;
}

main().catch((e) => {
  console.error(`Fatal error: ${e.message}`);
  process.exit(1);
});

File: c/tester.c
// feather-c is the C host interpreter for the feather test harness.
// It links against libfeather.so (built from the Go package) and provides
// REPL, script mode, and benchmark mode.

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include "libfeather.h"

// Convenience typedefs matching the Go exports
typedef size_t FeatherInterp;
typedef size_t FeatherObj;

#define FEATHER_OK 0
#define FEATHER_PARSE_OK 0
#define FEATHER_PARSE_INCOMPLETE 1
#define FEATHER_PARSE_ERROR 2

// -----------------------------------------------------------------------------
// Utility function to copy string to stack buffer with null termination
// -----------------------------------------------------------------------------

static void copy_string(FeatherInterp interp, FeatherObj obj, char *buf, size_t bufsize) {
    size_t len = FeatherLen(interp, obj);
    if (len >= bufsize) {
        len = bufsize - 1;
    }
    FeatherCopy(interp, obj, buf, len);
    buf[len] = '\0';
}

// Helper to create error object from format string
static FeatherObj make_error(FeatherInterp interp, const char *fmt, ...) {
    char buf[512];
    va_list args;
    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);
    return FeatherString(interp, buf, strlen(buf));
}

// -----------------------------------------------------------------------------
// Harness Protocol
// -----------------------------------------------------------------------------

static int in_harness = 0;
static FILE *harness_fd = NULL;

static void write_harness_result(const char *code, const char *result, const char *error) {
    if (!in_harness || !harness_fd) return;

    fprintf(harness_fd, "return: %s\n", code);
    if (result && result[0]) {
        fprintf(harness_fd, "result: %s\n", result);
    }
    if (error && error[0]) {
        fprintf(harness_fd, "error: %s\n", error);
    }
    fflush(harness_fd);
}

// -----------------------------------------------------------------------------
// Test Commands (handle-based API with FeatherObj errors)
// -----------------------------------------------------------------------------

static int cmd_say_hello(void *userData, FeatherInterp interp, size_t argc, FeatherObj *argv,
                         FeatherObj *result, FeatherObj *error) {
    (void)userData; (void)argc; (void)argv; (void)error;
    printf("hello\n");
    *result = FeatherString(interp, "", 0);
    return 0;
}

static int cmd_echo(void *userData, FeatherInterp interp, size_t argc, FeatherObj *argv,
                    FeatherObj *result, FeatherObj *error) {
    (void)userData; (void)error;
    char buf[4096];
    for (size_t i = 0; i < argc; i++) {
        if (i > 0) printf(" ");
        copy_string(interp, argv[i], buf, sizeof(buf));
        printf("%s", buf);
    }
    printf("\n");
    *result = FeatherString(interp, "", 0);
    return 0;
}

static int cmd_count(void *userData, FeatherInterp interp, size_t argc, FeatherObj *argv,
                     FeatherObj *result, FeatherObj *error) {
    (void)userData; (void)argv; (void)error;
    // Return the count as an integer
    *result = FeatherInt(interp, (int64_t)argc);
    return 0;
}

static int cmd_list(void *userData, FeatherInterp interp, size_t argc, FeatherObj *argv,
                    FeatherObj *result, FeatherObj *error) {
    (void)userData; (void)error;

    // Build the list directly using FeatherList
    *result = FeatherList(interp, argc, argv);
    return 0;
}

// -----------------------------------------------------------------------------
// Counter Foreign Type
// -----------------------------------------------------------------------------

typedef struct {
    int value;
} Counter;

static void* counter_new(void *userData) {
    (void)userData;
    Counter *c = malloc(sizeof(Counter));
    if (c) c->value = 0;
    return c;
}

static int counter_invoke(void *instance, FeatherInterp interp, const char *method,
                          size_t argc, FeatherObj *argv,
                          FeatherObj *result, FeatherObj *error) {
    Counter *c = (Counter*)instance;

    if (strcmp(method, "get") == 0) {
        if (argc != 0) {
            *error = make_error(interp, "wrong # args: expected 0, got %zu", argc);
            return 1;
        }
        *result = FeatherInt(interp, (int64_t)c->value);
        return 0;
    }

    if (strcmp(method, "set") == 0) {
        if (argc != 1) {
            *error = make_error(interp, "wrong # args: expected 1, got %zu", argc);
            return 1;
        }
        int64_t val = FeatherAsInt(interp, argv[0], -999999);
        if (val == -999999) {
            // Check if it really was -999999 or a conversion failure
            char buf[256];
            copy_string(interp, argv[0], buf, sizeof(buf));
            // Simple check: if the string is "-999999", it's valid
            if (strcmp(buf, "-999999") != 0) {
                *error = make_error(interp, "argument 1: expected integer but got \"%s\"", buf);
                return 1;
            }
        }
        c->value = (int)val;
        *result = FeatherString(interp, "", 0);
        return 0;
    }

    if (strcmp(method, "incr") == 0) {
        if (argc != 0) {
            *error = make_error(interp, "wrong # args: expected 0, got %zu", argc);
            return 1;
        }
        c->value++;
        *result = FeatherInt(interp, (int64_t)c->value);
        return 0;
    }

    if (strcmp(method, "add") == 0) {
        if (argc != 1) {
            *error = make_error(interp, "wrong # args: expected 1, got %zu", argc);
            return 1;
        }
        int64_t val = FeatherAsInt(interp, argv[0], -999999);
        if (val == -999999) {
            char buf[256];
            copy_string(interp, argv[0], buf, sizeof(buf));
            if (strcmp(buf, "-999999") != 0) {
                *error = make_error(interp, "argument 1: expected integer but got \"%s\"", buf);
                return 1;
            }
        }
        c->value += (int)val;
        *result = FeatherInt(interp, (int64_t)c->value);
        return 0;
    }

    *error = make_error(interp, "unknown method \"%s\": must be get, set, incr, add, destroy", method);
    return 1;
}

static void counter_destroy(void *instance) {
    free(instance);
}

// -----------------------------------------------------------------------------
// Register Commands
// -----------------------------------------------------------------------------

static void register_test_commands(FeatherInterp interp) {
    // Set milestone variables (like Go version)
    FeatherObj milestone = FeatherString(interp, "m1", 2);
    FeatherSetVar(interp, "milestone", milestone);
    FeatherSetVar(interp, "current-step", milestone);

    // Register test commands
    FeatherRegister(interp, "say-hello", cmd_say_hello, NULL);
    FeatherRegister(interp, "echo", cmd_echo, NULL);
    FeatherRegister(interp, "count", cmd_count, NULL);
    FeatherRegister(interp, "list", cmd_list, NULL);

    // Register Counter type
    FeatherRegisterForeign(interp, "Counter", counter_new, counter_invoke, counter_destroy, NULL);

    // Register Counter methods for info methods to work
    FeatherRegisterForeignMethod(interp, "Counter", "get");
    FeatherRegisterForeignMethod(interp, "Counter", "set");
    FeatherRegisterForeignMethod(interp, "Counter", "incr");
    FeatherRegisterForeignMethod(interp, "Counter", "add");
}

// -----------------------------------------------------------------------------
// REPL Mode
// -----------------------------------------------------------------------------

static void run_repl(FeatherInterp interp) {
    char line[4096];
    char input_buffer[65536];
    char result_buf[4096];
    input_buffer[0] = '\0';

    for (;;) {
        if (input_buffer[0] == '\0') {
            printf("%% ");
        } else {
            printf("> ");
        }
        fflush(stdout);

        if (!fgets(line, sizeof(line), stdin)) {
            break;
        }

        // Remove trailing newline
        size_t len = strlen(line);
        if (len > 0 && line[len-1] == '\n') {
            line[len-1] = '\0';
        }

        // Append to input buffer
        if (input_buffer[0] != '\0') {
            strcat(input_buffer, "\n");
        }
        strcat(input_buffer, line);

        // Evaluate
        FeatherObj result;
        int status = FeatherEval(interp, input_buffer, strlen(input_buffer), &result);

        if (status != FEATHER_OK) {
            copy_string(interp, result, result_buf, sizeof(result_buf));
            fprintf(stderr, "error: %s\n", result_buf);
        } else {
            copy_string(interp, result, result_buf, sizeof(result_buf));
            if (result_buf[0]) {
                printf("%s\n", result_buf);
            }
        }

        input_buffer[0] = '\0';
    }
}

// -----------------------------------------------------------------------------
// Script Mode
// -----------------------------------------------------------------------------

static char* read_all_stdin(size_t *out_len) {
    size_t capacity = 4096;
    size_t len = 0;
    char *buf = malloc(capacity);
    if (!buf) return NULL;

    for (;;) {
        if (len + 1024 > capacity) {
            capacity *= 2;
            char *new_buf = realloc(buf, capacity);
            if (!new_buf) {
                free(buf);
                return NULL;
            }
            buf = new_buf;
        }

        size_t n = fread(buf + len, 1, capacity - len - 1, stdin);
        if (n == 0) break;
        len += n;
    }

    buf[len] = '\0';
    *out_len = len;
    return buf;
}

static void run_script(FeatherInterp interp) {
    size_t script_len;
    char *script = read_all_stdin(&script_len);
    if (!script) {
        fprintf(stderr, "error reading script\n");
        write_harness_result("TCL_ERROR", "", "error reading script");
        exit(1);
    }

    // Check parse status first (like Go version)
    FeatherObj parseResultObj = 0;
    FeatherObj parseErrorObj = 0;
    int parseStatus = FeatherParseInfo(interp, script, script_len, &parseResultObj, &parseErrorObj);

    char result_buf[65536];

    if (parseStatus == FEATHER_PARSE_INCOMPLETE) {
        // Return parse result as TCL_OK with the INCOMPLETE info
        copy_string(interp, parseResultObj, result_buf, sizeof(result_buf));
        write_harness_result("TCL_OK", result_buf, "");
        free(script);
        exit(2);
    }

    if (parseStatus == FEATHER_PARSE_ERROR) {
        // Return parse result with error info
        copy_string(interp, parseResultObj, result_buf, sizeof(result_buf));
        char error_buf[4096] = "";
        if (parseErrorObj) {
            copy_string(interp, parseErrorObj, error_buf, sizeof(error_buf));
        }
        write_harness_result("TCL_ERROR", result_buf, error_buf);
        free(script);
        exit(3);
    }

    // Parse OK - evaluate the script
    FeatherObj result;
    int status = FeatherEval(interp, script, script_len, &result);
    free(script);

    copy_string(interp, result, result_buf, sizeof(result_buf));

    if (status != FEATHER_OK) {
        if (result_buf[0]) {
            printf("%s\n", result_buf);
        }
        write_harness_result("TCL_ERROR", "", result_buf);
        exit(1);
    }

    if (result_buf[0]) {
        printf("%s\n", result_buf);
    }
    write_harness_result("TCL_OK", result_buf, "");
}

// -----------------------------------------------------------------------------
// Benchmark Mode (simplified - no JSON parsing)
// -----------------------------------------------------------------------------

static void run_benchmark_mode(void) {
    // For now, just print an error - full JSON support requires more code
    fprintf(stderr, "benchmark mode not yet implemented\n");
    exit(1);
}

// -----------------------------------------------------------------------------
// Main
// -----------------------------------------------------------------------------

int main(int argc, char *argv[]) {
    // Check for benchmark mode
    if (argc > 1 && strcmp(argv[1], "--benchmark") == 0) {
        run_benchmark_mode();
        return 0;
    }

    // Check harness mode
    const char *harness_env = getenv("FEATHER_IN_HARNESS");
    if (harness_env && strcmp(harness_env, "1") == 0) {
        in_harness = 1;
        harness_fd = fdopen(3, "w");
    }

    // Create interpreter
    FeatherInterp interp = FeatherNew();
    register_test_commands(interp);

    // Check if stdin is TTY
    if (isatty(0)) {
        run_repl(interp);
    } else {
        run_script(interp);
    }

    FeatherClose(interp);
    return 0;
}
